src/linker/Linker/MessageContainer.cs

```diff
@@ -0,0 +1,105 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Text;
+
+namespace Mono.Linker
+{
+	public readonly struct MessageContainer
+	{
+		/// <summary>
+		/// Optional data with a filename, line and column that triggered the
+		/// linker to output an error (or warning) message.
+		/// </summary>
+		public MessageOrigin? Origin { get; }
+
+		public MessageCategory Category { get; }
+
+		/// <summary>
+		/// Further categorize the message.
+		/// </summary>
+		public string SubCategory { get; }
+
+		/// <summary>
+		/// Code identifier for errors and warnings reported by the IL linker.
+		/// </summary>
+		public int Code { get; }
+
+		/// <summary>
+		/// Optional user friendly text describing the error or warning.
+		/// </summary>
+		public string Text { get; }
+
+		public MessageContainer (MessageCategory category, string text, int code, string subcategory = MessageSubCategory.None, MessageOrigin? origin = null)
+		{
+			switch (code) {
+				// Errors
+				case int _code when (_code > 0 && _code <= 2000):
+					if (category != MessageCategory.Error)
+						throw new ArgumentException ($"MSBuild Message Container with code ${code} was expected to be of 'Error' category.");
+					break;
+
+				// Warnings
+				case int _code when (_code > 2000 && _code <= 6000):
+					if (category != MessageCategory.Warning)
+						throw new ArgumentException ($"MSBuild Message Container with code ${code} was expected to be of 'Warning' category.");
+					break;
+
+				// Info.
+				case int _code when (_code > 6000 && _code <= 8000):
```

Comment:
Should we make code optional for info and diagnostic messages? It looks like they are never displayed on UI so the code is not really useful.

Reply:
Yes, there's no need for the code parameter for this category (If we are to keep the `info` category).

Reply:
Please change Code property to be nullable then. I am also thinking of having factory methods instead of this constructor if only some parameters are required. Something like

```c#
static MessageContainer CreateErrorMessage (string text, int code, string subcategory = MessageSubCategory.None, MessageOrigin? origin = null) { .. }

static MessageContainer CreateWarningMessage (string text, int code, string subcategory = MessageSubCategory.None, MessageOrigin? origin = null) { .. }

static MessageContainer CreateInfoMessage (string text) { .. }

static MessageContainer CreateDiagnostingMessage (string text) { .. }
```